modules/wallet/update.go

```diff
@@ -18,10 +18,60 @@ type historicOutput struct {
 	val types.Currency
 }
 
+// isFutureWalletAddress is a helper function that checks if an UnlockHash is
+// derived from one of the wallet's future keys. If it is all future keys up to
+// this one will be added to the spendable keys or a blockhain rescan will be
+// initiated
+func (w *Wallet) isFutureWalletAddress(uh types.UnlockHash) bool {
+	if _, exists := w.futureKeys[uh]; !exists {
+		return false
+	}
+
+	// generate and add spendable keys until the future key is present in w.keys
+	foundKey := false
+	for i := 0; i < 1000; i++ {
+		_, err := w.nextPrimarySeedAddress(w.dbTx)
+		if err != nil {
+			return false
+		}
+
+		if _, exists := w.keys[uh]; exists {
+			foundKey = true
+			break
+		}
+	}
+
+	// if future key was not among the next 1000 keys rescan the blockchain
+	if !foundKey {
+		// TODO: rescan blockchain
+		/*
+			w.cs.Unsubscribe(w)
+			w.tpool.Unsubscribe(w)
+
+			done := make(chan struct{})
+			go w.rescanMessage(done)
+			defer close(done)
+
+			err := w.cs.ConsensusSetSubscribe(w, modules.ConsensusChangeBeginning)
+			if err != nil {
+				// TODO: error handling
+			}
+			w.tpool.TransactionPoolSubscribe(w)
+
+		*/
+		return false
+	}
+
+	return true
+}
+
 // isWalletAddress is a helper function that checks if an UnlockHash is
-// derived from one of the wallet's spendable keys.
+// derived from one of the wallet's spendable keys or future keys.
 func (w *Wallet) isWalletAddress(uh types.UnlockHash) bool {
 	_, exists := w.keys[uh]
+	if !exists {
+		exists = w.isFutureWalletAddress(uh)
+	}
```

Comment:
Is `isFutureWalletAddress` going to be useful on its own? If not, I would just inline it here:
```go
func (w *Wallet) isWalletAddress(uh types.UnlockHash) bool {
	_, exists := w.keys[uh]
	if exists {
		return true
	}
	// check future keys
	index, exists := w.futureKeys[uh]
	if exists {
		// advance the lookahead accordingly
		w.advanceSeedLookahead(index)
	}
	return exists
}
```
where `advanceSeedLookahead` adds all the keys in the range `currentSeedProgress -> index` to `w.keys`, deletes them from `w.futureKeys`, and adds all the keys in the range `(currentSeedProgress + lookahead) -> (index + lookahead)` to `w.futureKeys`.
I'm not sure how best to structure the lookahead stuff; it might be more readable to write something like `regenerateLookahead(min, max)` instead. But broadly speaking, I think "lookahead" communicates the purpose more directly than "future keys."